python/ql/lib/semmle/python/AstExtended.qll

```diff
@@ -220,3 +220,43 @@ class StringList extends StringList_ { }
 
 /** A list of aliases in an import statement */
 class AliasList extends AliasList_ { }
+
+/** A generic type parameter, as seen in function, class, and type alias definitions. */
+class TypeParameter extends TypeParameter_, AstNode {
+  /** Gets a textual representation of this element */
+  override string toString() { result = TypeParameter_.super.toString() }
+
+  override AstNode getAChildNode() { none() }
+
+  override Scope getScope() { none() }
```

Comment:
I think it looks strange that the getScope method has no result.

I investigated how `DictItem` works, which also [implements getScope as `none()`](https://github.com/github/codeql/blob/f67c68da9acfab5d705d9a47cdbc0e27f7dbdde7/python/ql/lib/semmle/python/AstExtended.qll#L107). However, I found that all its' subclasses provide an overriding getScope implementation:

* [KeyValuePair](https://github.com/github/codeql/blob/663096fe3a9b0ffd8eed1a3f0ec2eaddecc58971/python/ql/lib/semmle/python/Keywords.qll#L10)
* [DictUnpacking](https://github.com/github/codeql/blob/663096fe3a9b0ffd8eed1a3f0ec2eaddecc58971/python/ql/lib/semmle/python/Keywords.qll#L26)
* [Keyword](https://github.com/github/codeql/blob/663096fe3a9b0ffd8eed1a3f0ec2eaddecc58971/python/ql/lib/semmle/python/Keywords.qll#L51)

So any DictItem will actually have a valid .getScope() result due to this.

As far as I can tell, currently no other AstNode have empty .getScope, so I think it could make sense to add something here as well :blush:

I wasn't quite sure what the scope should be though :thinking: So I had a look at the [official Python grammar](https://github.com/python/cpython/blob/d61313bdb1eee3e4bb111e0b248ac2dbb48be917/Parser/Python.asdl#L11-L23). To me it seems reasonable that the scope should be the class/function the type-parameter is attached to, or the same scope as the containing `TypeAlias` -- what do you think?

Closing thought: I guess it's this is more of a nitpick than a _super important problem that will make everything break_, but it's just sort of strange that a AST node doesn't have any scope :thinking: :upside_down_face:

Reply:
Oh, you're completely right. Good catch!

The scoping of type variables is a bit peculiar in Python, so I'll have to think about what the right solution is. I agree that for `def foo[T](...)` and `class Foo[T]: ...`, the scope of `T` is the function/class itself (as this construction makes `T` available for the body of the function/class).

For `type Foo[T] = expr`, I feel like this creates a scope of its own, since it allows `T` to be referenced inside `expr`, but not outside the type alias. In practice, however, it's probably harmless to give it the same scope as the one containing the type alias.

Reply:
> For `type Foo[T] = expr`, I feel like this creates a scope of its own, since it allows `T` to be referenced inside `expr`, but not outside the type alias. In practice, however, it's probably harmless to give it the same scope as the one containing the type alias.

Ah, I had missed that part. In [What's new in Python 3.12](https://docs.python.org/3/whatsnew/3.12.html) they mention:

> In order to support these scoping semantics, a new kind of scope is introduced, the annotation scope. Annotation scopes behave for the most part like function scopes, but interact differently with enclosing class scopes. In Python 3.13, annotations will also be evaluated in annotation scopes.

So I guess a "proper" fix is to follow the scoping behaviors described here: https://peps.python.org/pep-0695/#scoping-behavior

I guess this calls for a test of whether we can resolve a reference to `T` inside `expr` -- I'm actually not 100% sure what the result would be, although I not super optimistic about it :grimacing:

Reply:
I thought about this for a bit, and the way I see it there are two options here.

Firstly, we can accept the current state (+ the scope fixes I'll be adding shortly). That is
- Type parameters on functions and classes are scoped to these entities.
- Type parameters on type aliases have the same scope as the alias itself (which is wrong, but harmless).
- There are no control-flow nodes for any of these new constructs (in line with the fact that they are not eagerly evaluated in Python).
- There's no data-flow for these values (because they don't ever have any values with which to flow).

Alternatively, we could add control-flow nodes for these constructs, and pretend that they do get evaluated at runtime (we can allow ourselves to be a bit more lax than the Python interpreter). In particular, we would want control-flow that evaluates type bounds before the body of the function/class/type alias, and also that the right hand side of a type alias gets evaluated before the next statement. We could also add flow steps from these parameters to the places where they are used (taking into account things like shadowing of type parameters).

The thing that's not clear to me is what we actually gain from this more precise modelling. The best I can come up with is that we'll (maybe?) have a harder time 
modelling things like
```python
def request_handler[R: Request](r: R):
    ...
```
where we have a handler that's generic in some overall type of requests, and we want to identify that `r` is indeed a value containing a request. But it also seems to me that this should be possible to model without the need of data-flow. In particular, it seems to me that these type parameters are only meant to be used within a given scope (that is, what would it even mean to pass a type parameter as an argument into a call to some other function?). If we really do want this option, though, I think we should also extend our modelling to support defining these things using the `typing` module (seeing as not a lot of code will be using Python 3.12 yet).

The two options above are not incompatible. We can easily do the first, and then revisit it if it turns out we need the second solution at some point.

Reply:
Getting the first set of fixes for Scope in now, and then optionally doing the second part later sounds good to me :+1: (I should probably have been more explicit about this).

If you can come up with a test whether we can resolve a reference to `T` inside `expr`, that would be a nice addition (just to document we know we don't handle this currently) -- Not going to make this a strict requirement though, I'm probably feeling like 3/10 on it :blush:

Reply:
Let's push it to a follow-up PR, as I'm starting to get submodulitis. (I'm also writing up an issue based on the above discussion, so we don't forget what to do in case we want to revisit it in the future.)